Kernel/Arch/riscv64/PageDirectory.cpp

```diff
@@ -8,39 +8,84 @@
 #include <AK/Singleton.h>
 
 #include <Kernel/Arch/PageDirectory.h>
+#include <Kernel/Interrupts/InterruptDisabler.h>
 #include <Kernel/Library/LockRefPtr.h>
 #include <Kernel/Sections.h>
+#include <Kernel/Tasks/Process.h>
+#include <Kernel/Tasks/Thread.h>
 
 namespace Kernel::Memory {
 
-void PageDirectory::register_page_directory(PageDirectory*)
+struct SATPMap {
+    SpinlockProtected<IntrusiveRedBlackTree<&PageDirectory::m_tree_node>, LockRank::None> map {};
+};
+
+static Singleton<SATPMap> s_satp_map;
+
+void PageDirectory::register_page_directory(PageDirectory* page_directory)
 {
-    TODO_RISCV64();
+    s_satp_map->map.with([&](auto& map) {
+        map.insert(bit_cast<FlatPtr>(page_directory->satp()), *page_directory);
+    });
 }
 
-void PageDirectory::deregister_page_directory(PageDirectory*)
+void PageDirectory::deregister_page_directory(PageDirectory* page_directory)
 {
-    TODO_RISCV64();
+    s_satp_map->map.with([&](auto& map) {
+        map.remove(bit_cast<FlatPtr>(page_directory->satp()));
+    });
 }
 
-ErrorOr<NonnullLockRefPtr<PageDirectory>> PageDirectory::try_create_for_userspace(Process&)
+ErrorOr<NonnullLockRefPtr<PageDirectory>> PageDirectory::try_create_for_userspace(Process& process)
 {
-    TODO_RISCV64();
+    auto directory = TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) PageDirectory));
+
+    directory->m_process = &process;
+
+    directory->m_directory_table = TRY(MM.allocate_physical_page());
+    auto kernel_pd_index = (kernel_mapping_base >> VPN_2_OFFSET) & PAGE_TABLE_INDEX_MASK;
+    for (size_t i = 0; i < kernel_pd_index; i++) {
+        directory->m_directory_pages[i] = TRY(MM.allocate_physical_page());
+    }
+
+    // Share the top 1 GiB of kernel-only mappings (>=kernel_mapping_base)
+    directory->m_directory_pages[kernel_pd_index] = MM.kernel_page_directory().m_directory_pages[kernel_pd_index];
+
+    {
+        InterruptDisabler disabler;
+        auto& table = *(PageDirectoryPointerTable*)MM.quickmap_page(*directory->m_directory_table);
```

Comment:
Would prefer a c++ cast here (`reinterpret_cast` in this case as both are pointers)

Reply:
I didn't know if I should change this while copy-pasting it from other arches.
But sure, I'll remove those C-style casts.